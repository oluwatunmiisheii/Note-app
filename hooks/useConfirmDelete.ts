import { Alert } from "react-native";

export const defaultConfirmDeleteText =
  "Are you sure you want to delete this note?";

export default function useConfirmDelete() {
  const confirmDelete = (
    onConfirm: () => void,
    text = defaultConfirmDeleteText
  ) => {
    Alert.alert(
      "Delete",
      text,
      [
        {
          text: "Cancel",
          style: "cancel",
        },
        { text: "Delete", onPress: onConfirm },
      ],
      { cancelable: false }
    );
  };

  return confirmDelete;
}
