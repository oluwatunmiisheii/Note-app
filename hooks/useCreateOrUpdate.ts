import { Alert } from "react-native";
import { useGlobalContext } from "../context/Global.context";

interface CreateOrUpdateProps<T> {
  fn: (...args: any) => Promise<T>;
  args: any;
  successCb?: (payload: Awaited<T>) => void;
  errorCb?: (error: unknown) => void;
  successMessage?: string;
  errorMessage?: string;
}

export const useCreateOrUpdate = () => {
  const { setLoading } = useGlobalContext();

  const createOrUpdate = async <T>({
    fn,
    args,
    successCb,
    errorCb,
    successMessage,
    errorMessage,
  }: CreateOrUpdateProps<T>) => {
    setLoading(true);
    try {
      const res = await fn(...args);
      successMessage && Alert.alert("Success", successMessage);
      successCb?.(res);
    } catch (error) {
      errorMessage && Alert.alert("Error", errorMessage);
      errorCb?.(error);
      return error;
    } finally {
      setLoading(false);
    }
  };

  return createOrUpdate;
};
