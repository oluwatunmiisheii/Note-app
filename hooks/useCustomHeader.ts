import { useNavigation } from "@react-navigation/native";
import { useLayoutEffect } from "react";

type Args = {
  headerRight?: React.ReactNode;
  headerLeft?: React.ReactNode;
};

export default function useCustomHeader(args: Args) {
  const navigation = useNavigation();

  useLayoutEffect(() => {
    navigation.setOptions({
      headerRight: () => args.headerRight,
      headerLeft: () => args.headerLeft,
    });
  }, [navigation]);
}
