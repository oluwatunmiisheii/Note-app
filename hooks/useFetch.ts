import { useFocusEffect } from "@react-navigation/native";
import { useCallback, useState } from "react";

type Dependencies = Parameters<typeof useCallback>[1];

export default function useFetch<T>(
  callback: (...args: any) => Promise<T>,
  dependencies: Dependencies
) {
  const [data, setData] = useState<T | null>(null);
  const [error, setError] = useState<Error | null>(null);
  const [isLoading, setLoading] = useState(false);

  useFocusEffect(
    useCallback(() => {
      let isMounted = true;
      const fetchData = async () => {
        setLoading(true);
        try {
          const data = await callback();
          setData(data);
        } catch (error) {
          setError(error as Error);
        } finally {
          setLoading(false);
        }
      };

      fetchData();

      return () => {
        isMounted = false;
      };
    }, dependencies)
  );

  return { data, error, isLoading, setData };
}
