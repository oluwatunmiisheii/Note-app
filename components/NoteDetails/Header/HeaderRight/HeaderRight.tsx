import { Feather } from "@expo/vector-icons";
import { Pressable, View, StyleSheet } from "react-native";
import useConfirmDelete from "hooks/useConfirmDelete";

interface HeaderRightProps {
  onDelete: () => void;
  onEdit: () => void;
}

export const HeaderRight = ({ onDelete, onEdit }: HeaderRightProps) => {
  const confirmDelete = useConfirmDelete();

  return (
    <View style={styles.container}>
      <Pressable
        style={{ marginRight: 20 }}
        onPress={onEdit}
        accessibilityLabel="edit note"
      >
        <Feather name="edit" color="white" size={19} />
      </Pressable>
      <Pressable
        onPress={() => {
          confirmDelete(onDelete);
        }}
        accessibilityLabel="delete note"
      >
        <Feather name="trash-2" color="white" size={19} />
      </Pressable>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: "row",
    alignItems: "center",
  },
});
