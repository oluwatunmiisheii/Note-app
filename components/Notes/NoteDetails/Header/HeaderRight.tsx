import { AntDesign, Feather } from "@expo/vector-icons";
import { Pressable, View } from "react-native";
import useConfirmDelete from "../../../../hooks/useConfirmDelete";

interface HeaderRightProps {
  onDelete: () => void;
  onEdit: () => void;
}

export const HeaderRight = ({ onDelete, onEdit }: HeaderRightProps) => {
  const confirmDelete = useConfirmDelete();

  return (
    <View style={{ flexDirection: "row", alignItems: "center" }}>
      <Pressable style={{ marginRight: 20 }} onPress={onEdit}>
        <AntDesign name="edit" color="white" size={24} />
      </Pressable>
      <Pressable
        onPress={() => {
          confirmDelete(onDelete);
        }}
      >
        <Feather name="trash-2" color="white" size={24} />
      </Pressable>
    </View>
  );
};
