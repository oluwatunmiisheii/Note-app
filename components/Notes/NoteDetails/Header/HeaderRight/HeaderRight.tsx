import { AntDesign, Feather } from "@expo/vector-icons";
import { Pressable, View } from "react-native";
import useConfirmDelete from "../../../../../hooks/useConfirmDelete";

interface HeaderRightProps {
  onDelete: () => void;
  onEdit: () => void;
}

export const HeaderRight = ({ onDelete, onEdit }: HeaderRightProps) => {
  const confirmDelete = useConfirmDelete();

  return (
    <View style={{ flexDirection: "row", alignItems: "center" }}>
      <Pressable
        style={{ marginRight: 20 }}
        onPress={onEdit}
        accessibilityLabel="edit note"
      >
        <AntDesign name="edit" color="white" size={18} />
      </Pressable>
      <Pressable
        onPress={() => {
          confirmDelete(onDelete);
        }}
        accessibilityLabel="delete note"
      >
        <Feather name="trash-2" color="white" size={18} />
      </Pressable>
    </View>
  );
};
