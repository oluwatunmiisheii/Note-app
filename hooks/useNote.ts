import { CreateNoteDto, Note } from "@/models/note.model";
import notesService from "@/services/note/notes.service";
import { useCreateOrUpdate } from "./useCreateOrUpdate";

type CreateNoteArgs = {
  payload: CreateNoteDto;
  onSuccess: (res: { name: string }) => void;
};

type UpdateNoteArgs = {
  payload: CreateNoteDto;
  noteId: string;
  onSuccess: (res: Note) => void;
};

type DeleteNoteArgs = {
  noteId: string;
  onSuccess: () => void;
};

export const useNote = () => {
  const createOrUpdate = useCreateOrUpdate();

  const createNote = async ({
    payload,
    onSuccess,
  }: CreateNoteArgs): Promise<void> => {
    createOrUpdate({
      fn: notesService.createNote,
      args: [payload],
      successCb: (res) => {
        onSuccess(res);
      },
      successMessage: "Note created successfully",
      errorMessage: "Something went wrong",
    });
  };

  const updateNote = async ({
    payload,
    noteId,
    onSuccess,
  }: UpdateNoteArgs): Promise<void> => {
    createOrUpdate({
      fn: notesService.updateNoteById,
      args: [noteId, payload],
      successCb: (res) => {
        onSuccess({
          ...res,
          id: noteId,
          date: new Date(res.date),
        });
      },
      successMessage: "Note updated successfully",
      errorMessage: "Something went wrong",
    });
  };

  const deleteNote = async ({ noteId, onSuccess }: DeleteNoteArgs) => {
    createOrUpdate({
      fn: notesService.deleteNoteById,
      args: [noteId],
      successCb: onSuccess,
      successMessage: "Note deleted successfully",
      errorMessage: "Something went wrong",
    });
  };

  return {
    createNote,
    updateNote,
    deleteNote,
  };
};
