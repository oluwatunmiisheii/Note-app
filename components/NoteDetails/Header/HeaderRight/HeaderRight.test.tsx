import renderer from "react-test-renderer";
import { render, screen, fireEvent } from "@testing-library/react-native";
import { HeaderRight } from "./HeaderRight";
import { Alert, AlertButton } from "react-native";
import { defaultConfirmDeleteText } from "../../../../hooks/useConfirmDelete";

const mockOnEdit = jest.fn();
const mockOnDelete = jest.fn();

const defaultProps: React.ComponentProps<typeof HeaderRight> = {
  onEdit: mockOnEdit,
  onDelete: mockOnDelete,
};

const create = () => {
  const tree = renderer.create(<HeaderRight {...defaultProps} />).toJSON();
  return { tree };
};

const setup = () => {
  const utils = render(<HeaderRight {...defaultProps} />);
  return { ...utils };
};

describe(`<${HeaderRight.name}>`, () => {
  test("renders correctly", () => {
    const { tree } = create();
    expect(tree).toMatchSnapshot();
  });

  test("should call onEdit on edit button press", () => {
    setup();
    const trigger = screen.getByLabelText("edit note");
    fireEvent(trigger, "press");
    expect(mockOnEdit).toBeCalledTimes(1);
  });

  test("should call onDelete on delete button press", async () => {
    const spy = jest.spyOn(Alert, "alert");
    setup();

    const trigger = screen.getByLabelText("delete note");
    fireEvent(trigger, "press");
    expect(spy).toBeCalledTimes(1);

    const [title, message, buttons] = spy.mock.calls[0];

    expect(title).toBe("Delete");
    expect(message).toBe(defaultConfirmDeleteText);
    expect(buttons).toHaveLength(2);

    const [cancelButton, deleteButton] = buttons as AlertButton[];

    expect(cancelButton.text).toBe("Cancel");
    expect(deleteButton.text).toBe("Delete");

    // @ts-ignore
    deleteButton.onPress();

    expect(mockOnDelete).toBeCalledTimes(1);
  });
});
