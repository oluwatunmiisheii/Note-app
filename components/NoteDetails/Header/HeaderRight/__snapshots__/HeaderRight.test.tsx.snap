// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HeaderRight> renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
    }
  }
>
  <View
    accessibilityLabel="edit note"
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "marginRight": 20,
      }
    }
  >
    <FeatherIcon
      color="white"
      name="edit"
      size={19}
    />
  </View>
  <View
    accessibilityLabel="delete note"
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <FeatherIcon
      color="white"
      name="trash-2"
      size={19}
    />
  </View>
</View>
`;
